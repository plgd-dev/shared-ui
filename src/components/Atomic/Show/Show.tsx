import { Children, ReactNode, ReactElement } from 'react'
import { Props, ComponentType } from './Show.types'

const Show: ComponentType<Props> = (props) => {
    let when: ReactNode | null = null
    let otherwise: ReactNode | null = null

    Children.forEach(props.children, (child) => {
        const children = child as ReactElement<{ isTrue?: boolean }>

        if (children.props.isTrue === undefined) {
            otherwise = children
        } else if (!when && children.props.isTrue === true) {
            when = children
        }
    })

    return when || otherwise
}

Show.displayName = 'Show'

Show.When = ({ isTrue, children }) => isTrue && children
Show.Else = ({ render, children }) => render || children

export default Show
