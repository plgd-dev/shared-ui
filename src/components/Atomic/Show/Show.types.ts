import { FC, ReactElement, ReactNode } from 'react'

export type Props = {
    children: ReactNode
}

export interface ComponentType<P> extends FC<P> {
    When: FC<WhenProps>
    Else: FC<ElseProps>
}

export type WhenProps = {
    isTrue: boolean
    children: any
}

export type ElseProps = {
    render: ReactNode
    children: any
}
