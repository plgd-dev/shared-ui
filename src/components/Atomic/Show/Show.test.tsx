import { render } from '@testing-library/react'
import Show from './Show'

describe('Show Component', () => {
    it('renders the "when" child when condition is true', () => {
        const { getByText, asFragment } = render(
            <Show>
                <Show.When isTrue={true}>When Content</Show.When>
                <Show.Else>Else Content</Show.Else>
            </Show>
        )
        expect(getByText('When Content')).toBeInTheDocument()
        expect(asFragment()).toMatchSnapshot()
        expect(() => getByText('Else Content')).toThrow()
    })

    it('renders the "else" child when condition is false', () => {
        const { getByText } = render(
            <Show>
                <Show.When isTrue={false}>When Content</Show.When>
                <Show.Else>Else Content</Show.Else>
            </Show>
        )
        expect(getByText('Else Content')).toBeInTheDocument()
        expect(() => getByText('When Content')).toThrow()
    })
})
