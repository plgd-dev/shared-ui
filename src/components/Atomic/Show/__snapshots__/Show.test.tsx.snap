// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Component renders the "when" child when condition is true 1`] = `
<DocumentFragment>
  When Content
</DocumentFragment>
`;
